controllers/openstackcluster_controller.go

```diff
@@ -590,6 +590,60 @@ func bastionHashHasChanged(computeHash string, clusterAnnotations map[string]str
 	return latestHash != computeHash
 }
 
+func reconcileLoadBalancerNetwork(openStackCluster *infrav1.OpenStackCluster, networkingService *networking.Service) error {
+	if openStackCluster.Spec.APIServerLoadBalancer != nil && openStackCluster.Spec.APIServerLoadBalancer.Network != nil {
+		lbNetOpts := filterconvert.NetworkFilterToListOpts(openStackCluster.Spec.APIServerLoadBalancer.Network)
+		lbNetList, err := networkingService.GetNetworksByFilter(&lbNetOpts)
+		if err != nil {
+			handleUpdateOSCError(openStackCluster, fmt.Errorf("failed to find loadbalancer network: %w", err))
+			return fmt.Errorf("failed to find network: %w", err)
+		}
+		if len(lbNetList) == 0 {
+			handleUpdateOSCError(openStackCluster, fmt.Errorf("failed to find loadbalancer network"))
+			return fmt.Errorf("failed to find any network")
+		}
+		if len(lbNetList) > 1 {
+			handleUpdateOSCError(openStackCluster, fmt.Errorf("found multiple loadbalancer networks (result: %v)", lbNetList))
+			return fmt.Errorf("found multiple loadbalancer networks (result: %v)", lbNetList)
+		}
+
+		// Filter out only relevant subnets specified by the spec
+		if openStackCluster.Status.APIServerLoadBalancer == nil {
+			openStackCluster.Status.APIServerLoadBalancer = &infrav1.LoadBalancer{
+				LoadBalancerNetwork: &infrav1.NetworkStatusWithSubnets{
+					NetworkStatus: infrav1.NetworkStatus{
+						Name: lbNetList[0].Name,
+						ID:   lbNetList[0].ID,
+						Tags: lbNetList[0].Tags,
+					},
+				},
+			}
+		}
+
+		openStackCluster.Status.APIServerLoadBalancer.LoadBalancerNetwork.Subnets = []infrav1.Subnet{}
+		for _, s := range openStackCluster.Spec.APIServerLoadBalancer.Subnets {
+			matchFound := false
+			for _, subnetID := range lbNetList[0].Subnets {
+				if subnetID == s.ID {
+					matchFound = true
+					openStackCluster.Status.APIServerLoadBalancer.LoadBalancerNetwork.Subnets = append(
+						openStackCluster.Status.APIServerLoadBalancer.LoadBalancerNetwork.Subnets, infrav1.Subnet{
+							ID:   s.ID,
+							Name: s.Name,
+							CIDR: s.CIDR,
+						})
+				}
+			}
+			if !matchFound {
+				handleUpdateOSCError(openStackCluster, fmt.Errorf("no subnet match was found in the specified network (specified subnet: %v, available subnets: %v)", s, lbNetList[0].Subnets))
+				return fmt.Errorf("no subnet match was found in the specified network (specified subnet: %v, available subnets: %v)", s, lbNetList[0].Subnets)
+			}
+		}
+	}
```

Comment:
How about making the default explicit and copying in the cluster network in the else branch? Then we could just make the creation code assume it's set. We could also shortcut the lookup at the top of this function if it's already set.

Reply:
I think it looks better now.